crates/publish/src/error.rs

```diff
@@ -0,0 +1,40 @@
+/// A specialized Result type for publish operations
+pub type PublishResult<T> = std::result::Result<T, Error>;
+
+/// Describes various errors that can be returned during publishing
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// Failed because the path terminates in a root or prefix
+    #[error("Failed to get containing directory for app file '{0}'")]
+    AppDir(String),
+    /// The bindle already exists
+    #[error("Bindle {0} already exists on the server")]
+    BindleAlreadyExists(String),
+    /// Bindle client failure
+    #[error("Error creating bindle client")]
+    BindleClient(#[from] bindle::client::ClientError),
+    /// Malformed bindle id
+    #[error("App name and version '{0}' do not form a bindle ID")]
+    BindleId(String),
+    /// IO error happened during building bindle component manifest
+    #[error("Failed to build bindle component manifest: {path}")]
+    BindleComponentManifest {
+        /// Represents path on a filesystem
+        path: String,
+        /// Underlying lower level error that caused your error
+        source: std::io::Error,
+    },
+    /// Publishing of components whose sources are already bindles is not supported
+    #[error("This version of Spin can't publish components whose sources are already bindles")]
+    BindlePushingNotImplemented,
+    /// IO errors from interacting with the file system
+    #[error("Error while performing IO operation")]
+    Io(#[from] std::io::Error),
+    /// Invalid TOML serialization that can occur when serializing an object to a request
+    #[error("Failed to write app manifest to TOML")]
```

Comment:
Is the "app manifest" the same as the "bindle component manifest"?

Reply:
I think this description isn't correct. Will update to something more generic like "Failed to serialize to TOML" as it's also used to write `invoice.toml`.

Reply:
Sounds good.  If there's any way we can say _what_ we failed to serialise (the invoice, the app manifest, anything else?) that would be super helpful for diagnostics.

Reply:
done